java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java

```diff
@@ -0,0 +1,155 @@
+package com.datasqrl.ai;
+
+import com.datasqrl.ai.api.APIExecutor;
+import com.datasqrl.ai.api.APIExecutorFactory;
+import com.datasqrl.ai.api.RESTExecutorFactory;
+import com.datasqrl.ai.models.ChatSession;
+import com.datasqrl.ai.models.ContextWindow;
+import com.datasqrl.ai.models.openai.OpenAIModelBindings;
+import com.datasqrl.ai.models.openai.OpenAIModelConfiguration;
+import com.datasqrl.ai.tool.FunctionValidation;
+import com.datasqrl.ai.tool.RuntimeFunctionDefinition;
+import com.datasqrl.ai.tool.ToolsBackend;
+import com.datasqrl.ai.tool.ToolsBackendFactory;
+import com.theokanning.openai.completion.chat.AssistantMessage;
+import com.theokanning.openai.completion.chat.ChatCompletionChunk;
+import com.theokanning.openai.completion.chat.ChatCompletionRequest;
+import com.theokanning.openai.completion.chat.ChatFunctionCall;
+import com.theokanning.openai.completion.chat.ChatMessage;
+import com.theokanning.openai.completion.chat.FunctionMessage;
+import com.theokanning.openai.completion.chat.UserMessage;
+import com.theokanning.openai.service.OpenAiService;
+import io.reactivex.Flowable;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.configuration2.MapConfiguration;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A simple streaming chatbot for the command line.
+ */
+@Slf4j
+@Value
+public class LowLevelAgent {
+
+  String systemPrompt;
+  OpenAiService service;
+  ToolsBackend toolsBackend;
+  OpenAIModelConfiguration chatConfig = new OpenAIModelConfiguration(
+      new MapConfiguration(Map.of(
+          "name", "gpt-3.5-turbo",
+          "temperature", 0.3)));
+
+  /**
+   * Initializes a command line chat bot using OpenAI
+   *
+   * @param toolsPath     A path to a tools file in JSON format
+   * @param apiExecutor   An APIExecutor for the tools
+   * @param systemPrompt  The system message for the LLM
+   * @param openAIKey     The OpenAI API key to call the API
+   */
+
+  public LowLevelAgent(Path toolsPath, APIExecutor apiExecutor, String systemPrompt, String openAIKey) throws IOException {
```

Comment:
@mbroecheler Pls review this low level agent code carefully. I am not sure that I got the abstraction level right. Should I use the backend directly to validate the function call? Should I introduce the retry logic here? Or anything else that should be at a different abstraction level?

Reply:
I like that this uses the streaming API (even though that makes the code more complex it highlights that you have that low level control).
My suggestions:
1. Create a class MyChatSession that extends ChatSession to overwrite the `getContextWindow(int maxTokens, ModelAnalyzer<Message> analyzer)` function with custom context window creation logic. I think that's a piece a lot of people want fine grained control over (since it determines the prompt and hence the cost & quality). Do a simple implementation that only ever takes up to the last 2 messages into the context.
2. Add a comment in there saying that you'd likely want to add retry logic for failed function calls and point to ChatProvider implementation that does this (no need to copy the code here).